nixos/modules/programs/dconf.nix

```diff
@@ -1,55 +1,245 @@
 { config, lib, pkgs, ... }:
 
-with lib;
-
 let
   cfg = config.programs.dconf;
-  cfgDir = pkgs.symlinkJoin {
-    name = "dconf-system-config";
-    paths = map (x: "${x}/etc/dconf") cfg.packages;
-    postBuild = ''
-      mkdir -p $out/profile
-      mkdir -p $out/db
-    '' + (
-      concatStringsSep "\n" (
-        mapAttrsToList (
-          name: path: ''
-            ln -s ${path} $out/profile/${name}
-          ''
-        ) cfg.profiles
+
+  # Compile keyfiles to dconf DB
+  compileDconfDb = dir: pkgs.runCommand "dconf-db"
+    {
+      nativeBuildInputs = [ (lib.getBin pkgs.dconf) ];
+    } "dconf compile $out ${dir}";
+
+  # Recursively lock values in settings
+  recursiveLock =
+    settings: locks:
+    (lib.mapAttrsRecursive
+      (path: value:
+      lib.optionalAttrs value
+        (lib.mapAttrsRecursiveCond
+          (value': !lib.gvariant.isGVariant value')
+          (path': value': true)
+          (lib.foldl'
+            (root: name: if lib.hasAttr "${name}" root then root."${name}" else
+            throw "${lib.concatStringsSep "." (lib.tail path)} doesn't exist in settings.")
+            { v = settings; }
+            path)
+        )
       )
-    ) + ''
-      ${pkgs.dconf}/bin/dconf update $out/db
-    '';
+      { v = locks; }).v;
+
+  # Check if dconf keyfiles are valid
+  checkDconfKeyfiles = dir: pkgs.runCommand "check-dconf-keyfiles"
+    {
+      nativeBuildInputs = [ (lib.getBin pkgs.dconf) ];
+    } ''
+    if [[ -f ${dir} ]]; then
+      echo "Dconf keyfiles should be a directory but a file is provided: ${dir}"
+      exit 1
+    fi
+
+    dconf compile db ${dir} || (
+      echo "The Dconf keyfiles are invalid: ${dir}"
+      exit 1
+    )
+    cp -R ${dir} $out
+  '';
+
+  # Generate dconf DB from dconfDatabase and keyfiles
+  mkDconfDb = val: compileDconfDb (pkgs.symlinkJoin {
+    name = "nixos-generated-dconf-keyfiles";
+    paths = [
+      (pkgs.writeTextDir "nixos-generated-dconf-keyfiles" (lib.generators.toDconfINI val.settings))
+      (pkgs.writeTextDir "locks/nixos-generated-dconf-locks" (lib.generators.toDconfLocks
+        (lib.recursiveUpdate (recursiveLock val.settings val.lockSubpaths) val.locks)))
+    ] ++ (map checkDconfKeyfiles val.keyfiles);
+  });
+
+  # Check if a dconf DB file is valid. The dconf cli doesn't return 1 when it can't
+  # open the database file so we have to check if the output is empty.
+  checkDconfDb = file: pkgs.runCommand "check-dconf-db"
+    {
+      nativeBuildInputs = [ (lib.getBin pkgs.dconf) ];
+    } ''
+    if [[ -d ${file} ]]; then
+      echo "Dconf DB should be a file but a directory is provided: ${file}"
+      exit 1
+    fi
+
+    echo "file-db:${file}" > profile
+    DCONF_PROFILE=$(pwd)/profile dconf dump / > dump 2> error
+    if [[ ! -s output ]] && [[ -s error ]]; then
```

Comment:
Empty databases are apparently possible

```ShellSession
$ cd "$(mktemp -d)"
$ mkdir -p foo.d
$ touch foo.d/foo
$ dconf compile foo foo.d
$ echo file-db:foo > profile
$ DCONF_PROFILE=$PWD/profile dconf dump /
```

Not sure if this edge case is worth handling, though.

Reply:
Yes, the database can be empty but looks like it makes no sense to add an empty database.

Reply:
And I also check that there is no error so this should have been covered.

Reply:
You are right. That should work.